File: practice-java/src/com/augustine/algo/MergeSort.java
package com.augustine.algo;

import java.util.Arrays;

public class MergeSort {
    public static void mergeSort(int[] array, int n) {
        if (n < 2) {
            return;
        }
        int mid = n / 2;
        int[] l = new int[mid];
        int[] r = new int[n - mid];

        for (int i = 0; i < mid; i++) {
            l[i] = array[i];
        }
        for (int i = mid; i < n; i++) {
            r[i - mid] = array[i];
        }
        mergeSort(l, mid);
        mergeSort(r, n - mid);

        merge(array, l, r, mid, n - mid);
    }

    public static void merge(int[] array, int[] l, int[] r, int left, int right) {
        int i = 0, j = 0, k = 0;
        while (i < left && j < right) {
            if (l[i] <= r[j]) {
                array[k++] = l[i++];
            } else {
                array[k++] = r[j++];
            }
        }
        while (i < left) {
            array[k++] = l[i++];
        }
        while (j < right) {
            array[k++] = r[j++];
        }
    }

    public static void main(String[] args) {
        int[] array = { 6, 2, 4, 5, 1, 9, 7, 3 };
        mergeSort(array, array.length);
        System.out.println(Arrays.toString(array));
    }
}

File: ds-python/algo/merge_sort.py
def mergeSort(array):
    n = len(array)
    if n < 2:
        return
    mid = n//2
    l = array[:mid]
    r = array[mid:]

    mergeSort(l)
    mergeSort(r)

    merge(array, l, r, len(l), len(r))


def merge(array, l, r, left, right):
    i = j = k = 0
    while i < left and j < right:
        if l[i] < r[j]:
            array[k] = l[i]
            i += 1
        else:
            array[k] = r[j]
            j += 1
        k += 1
        
    while i < left:
        array[k] = l[i]
        k += 1
        i += 1
    while j < right:
        array[k] = r[j]
        k += 1
        j += 1

arr = [12, 11, 13, 5, 6, 7]
mergeSort(arr)
print(arr)

    